### PR TITLE
feat(bandwidth_scheduler) - generate bandwidth requests based on receipts in outgoing buffers

### PR DESCRIPTION
**Pull request by jancionear**
To generate a bandwidth request we have to take a look at the outgoing buffer to some shard and request values that correspond to the sizes of receipts stored in the buffer. Ideally we would walk over all the individual receipts, but that would be pretty slow. Instead of that let's keep coarse metadata about receipts in each outgoing buffer and walk over the metadata instead of individual receipts. The metadata is much smaller than the whole buffer, so walking over it will be much faster.

The metadata looks like this:
```rust
/// Keeps metadata about receipts stored in outgoing buffer to some shard.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct OutgoingBufferMetadata {
    pub receipt_group_sizes: ReceiptGroupSizes,
}

pub struct ReceiptGroupSizesV0 {
    /// Receipts stored in the outgoing buffer, grouped into groups of consecutive receipts.
    /// Each group stores aggregate information about the receipts in the group.
    pub groups: VecDeque<ReceiptGroup>,
    /// A new receipt group is started when the size of the last group exceeds this threshold.
    /// All groups (except for the first and last one) have at least this size.
    pub group_size_threshold: u32,
}

/// Aggregate information about a group of consecutive receipts.
#[derive(Debug, Clone, Copy, PartialEq, Eq, BorshSerialize, BorshDeserialize, ProtocolSchema)]
pub struct ReceiptGroup {
    /// Total size of receipts in this group.
    /// Size is stored as u32 to keep the struct small.
    pub group_size: u32,
}
```

Consecutive receipts are grouped into groups. A new group is started when the size of the last group reaches 90kB (`group_size_threshold`). The first/last group in the list is updated when the first/last receipt is popped/added.
To generate a bandwidth request we walk over the sizes of the groups, not individual receipts. It's not as exact as individual receipts would be, but it should be good enough. Distance between `BandwidthRequestValues` is about 111kB, so having 90kB granularity should be fine.

The list of receipt groups for every buffer is stored as a single trie value. A trie value can be at most 4MB, so we have space for 1 million receipt groups. We'll never reach this size under normal circumstances, it would require us to have 90 GB of receipts stored in the outgoing buffer.

Implementing the metadata was a bit troublesome. I think ideally we would have an `OutgoingBuffer` struct which keeps all information about the receipts stored in the buffer, including the metadata. It'd have a safe interface (push/pop/iter) that ensures that metadata is always updated when a receipt is popped or pushed. I tried to do something like this, but the existing code strongly resists. Currently `OutgoingReceiptBuffer` is a &mut reference to `ShardsOutgoingReceiptBuffer`, the `OutgoingReceiptBuffer` doesn't keep any state by itself. This makes it hard to store some information there. The &mut reference also makes it difficult to do const operations that don't require mut, like reading the list of receipts and generating a bandwidth request. There's also the `pop_n` function which directly manipulates queue indices, which doesn't allow to update metadata.
In the end I gave up on the idea and decided to store the buffer metadatas directly in `ReceiptSink`. They're updated every time a receipt is buffered or removed, just like congestion info.

I had to refactor `ReceiptSink` a bit - I wanted to pass it to the function that finalizes chunk application, but it borrowed `outgoing_receipts` and `own_congestion_info`, which made it impossible. I changed it so that `ReceiptSink` owns these values, and now it can be passed to the function.

The metadatas work a bit differently from the `TrieQueue` - `TrieQueue` saves the updated indices to the trie after every update, but the metadata could be a bit larger, so I didn't want to do that. Instead the metadata is saved at the end of chunk application. AFAIU this should be fine, that's how we do it for promise/yield as well.

Protocol upgrade is handled using the `StateStoredReceipt` enum. I added a new version which keeps information about the version of metadata that was updated when this receipt was stored. Thanks to this we know if we should update the metadata when a receipt is removed from the buffer. We can also check if there are any receipts not included in the metadata by looking at the first receipt in the buffer.

@shreyan-gupta the metadata could be used to get the total size of receipts in outgoing buffer without iterating through all of the receipts, which I heard is needed for resharding. I could make the metadata into a separate feature and release it before bandwidth scheduler.

The change is untested for now, I'd like to get some feedback before writing all the tests.

The PR is split into commits for easier review.

### CODE REVIEW

**Comment by wacban** [comment]
cc @shreyan-gupta - Can you double check that we can easily do resharding no the metadata please?

File: core/primitives/src/trie_key.rs [hunk 1]
```diff
@@ -178,6 +180,9 @@ pub enum TrieKey {
         index: u64,
     },
     BandwidthSchedulerState,
+    OutgoingBufferReceiptSizes {
+        to_shard: ShardId,
+    },
```
Comment (wacban) [inline core/primitives/src/trie_key.rs:185]: You keep the entire structure in a single (per shard) trie value. This means that every access will need to retrieve and parse the entire thing adding to the state witness size. How large is it? If it is non-trivial I would suggest implementing a more fine grained storage like the `indices` + `values` structure for storing receipts in state.

Comment (wacban) [reply core/primitives/src/trie_key.rs:185]: nits: 
- The `BufferedReceipt` does not have the Outgoing prefix - I would apply the same convention here to make them more similar
- Maybe replace Sizes with groups to make it a bit more future proof? 
- How about `BufferedReceiptGroups`? Then we'll have `BufferedReceipt`, `BufferedReceiptIndices` and `BufferedReceiptGroups`.

File: core/store/src/trie/outgoing_metadata.rs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+/// Consecutive receipts are grouped into groups and it keeps aggregate information about the groups,
+/// not individual receipts. Receipts are grouped to keep the size of the struct small.
+#[derive(Debug, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize, ProtocolSchema)]
+pub struct ReceiptGroupSizesV0 {
+    /// Receipts stored in the outgoing buffer, grouped into groups of consecutive receipts.
+    /// Each group stores aggregate information about the receipts in the group.
+    pub groups: VecDeque<ReceiptGroup>,
+    /// A new receipt group is started when the size of the last group exceeds this threshold.
+    /// All groups (except for the first and last one) have at least this size.
+    pub group_size_threshold: u32,
```
Comment (wacban) [inline core/store/src/trie/outgoing_metadata.rs:138]: Does this need to be part of this structure? I suppose you keep it here to support future migrations. Can you explain what are the alternatives? In particular I'm curious about the pros and cons of storing all groups in a single trie node vs storing each in a dedicated trie node.

**Review by wacban — COMMENTED** [review]
Just a high level first iteration while I'm warming up to the idea

File: core/primitives/src/receipt.rs [hunk 3]
```diff
@@ -113,6 +115,24 @@ pub struct StateStoredReceiptV0<'a> {
     pub metadata: StateStoredReceiptMetadata,
 }
 
+/// The V1 of StateStoredReceipt.
+#[derive(BorshDeserialize, BorshSerialize, PartialEq, Eq, Debug, ProtocolSchema)]
+pub struct StateStoredReceiptV1<'a> {
+    pub receipt: Cow<'a, Receipt>,
+    pub metadata: StateStoredReceiptMetadata,
+    /// Version of buffer metadata that was updated when the receipt was stored.
+    pub queue_metadata_version: QueueMetadataVersion,
```
Comment (shreyan-gupta) [inline core/primitives/src/receipt.rs:124]: I don't quite get the need for this field. If we are already versioning `StateStoredReceipt`, each new metadata version can increment StateStoredReceipt version or we can make StateStoredReceiptMetadata a versioned enum itself like Block and BlockHeaders.

File: core/primitives/src/receipt.rs [hunk 4]
```diff
@@ -113,6 +115,24 @@ pub struct StateStoredReceiptV0<'a> {
     pub metadata: StateStoredReceiptMetadata,
 }
 
+/// The V1 of StateStoredReceipt.
+#[derive(BorshDeserialize, BorshSerialize, PartialEq, Eq, Debug, ProtocolSchema)]
+pub struct StateStoredReceiptV1<'a> {
+    pub receipt: Cow<'a, Receipt>,
```
Comment (shreyan-gupta) [inline core/primitives/src/receipt.rs:121]: @wacban, maybe a good time to cleanup the Cow part of the receipt here which you've been meaning to do.

**Review by shreyan-gupta — COMMENTED** [review]
Ok, I honestly don't get the design completely. Could you please point me to the design doc once again?

- What is the base requirement for sending a bandwidth request? What all pieces of information do we need?
- Is it just the size of the buffered receipts queue? In which case...
- Why are we not working with just storing one value (per shard ofc) in the CongestionInfo about the buffered receipt sizes?
- Why do we need this complex dequeue structure to be stored and updated as metadata?
- And why does this structure need to be part of trie and not something like CongestionInfo (in case it's small)?

File: core/primitives/src/bandwidth_scheduler.rs [hunk 5]
```diff
@@ -129,6 +129,39 @@ impl BandwidthRequest {
...
     }
+
+    /// Create a basic bandwidth request when receipt sizes are not available.
+    /// It'll request two values - one corresponding to the first receipt in
+    /// the outgoing buffer and one corresponding to max_receipt_size.
+    /// This ensures that the blockchain will be able to make progress while
+    /// waiting for the receipt sizes to become available.
+    /// The resulting behaviour is similar to the previous approach with allowed shard.
+    pub fn make_basic(
+        to_shard: u8,
```
[the remark this answers is not in the document]
Comment (jancionear) [reply core/primitives/src/bandwidth_scheduler.rs:140]: I was trying to save space, since the requests are in chunk header and we could have num_shards**2 of them in the worst case. Switching to u16 when the time comes shouldn't be too hard, `BandwidthRequests` is a versioned struct.
It would be nice to have something like "compact int representation" in borsh, e.g https://reference.cash/protocol/formats/variable-length-integer

File: core/primitives-core/src/types.rs [hunk 6]
```diff
@@ -183,6 +183,14 @@ impl Into<usize> for ShardId {
     }
 }
 
+impl TryInto<u8> for ShardId {
```
[the remark this answers is not in the document]
Comment (jancionear) [reply core/primitives-core/src/types.rs:186]: ShardId is stored as `u8` to save space in chunk headers, see https://github.com/near/nearcore/pull/12375#discussion_r1831129081

File: core/primitives/src/bandwidth_scheduler.rs [hunk 7]
```diff
@@ -129,6 +129,39 @@ impl BandwidthRequest {
...
+        }
+
+        for i in 0..values.len() {
+            if values[i] == params.max_receipt_size {
+                bitmap.set_bit(i, true);
+                break;
+            }
+        }
+
+        BandwidthRequest { to_shard, requested_values_bitmap: bitmap }
```
[the remark this answers is not in the document]
Comment (jancionear) [reply core/primitives/src/bandwidth_scheduler.rs:163]: There's a comment on `BandwidthRequestValues`, right below the bandwidth request struct.
```rust
/// There are this many predefined values of bandwidth that can be requested in a BandwidthRequest.
pub const BANDWIDTH_REQUEST_VALUES_NUM: usize = 40;

/// Values of bandwidth that can be requested in a bandwidth request.
/// When the nth bit is set in a request bitmap, it means that a shard is requesting the nth value from this list.
/// The list is sorted, from smallest to largest values.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct BandwidthRequestValues {
    pub values: [Bandwidth; BANDWIDTH_REQUEST_VALUES_NUM],
}
```
I could add a short comment which points to `BandwidthRequestValues` and `BandwidthRequestBitmap`

**Comment by jancionear** [comment]
> What is the base requirement for sending a bandwidth request? What all pieces of information do we need?

A bandwidth request means - "I have these receipts that I would like to send to the other shard, but sending them all might overload the receiver, can I be granted bandwidth to send (some of) them?".
For small amounts of bandwidth (less than base_bandwidth, about 50kB atm), there's no need to make a request, a shard can just send the data. For larger amounts it needs to make a request. 

> Is it just the size of the buffered receipts queue? In which case...
> Why are we not working with just storing one value (per shard ofc) in the CongestionInfo about the buffered receipt sizes?
> Why do we need this complex dequeue structure to be stored and updated as metadata?

The problem with having one value is that the bandwidth scheduler can't really grant partial bandwidth. When a shard wants to send one 4MB receipt, granting 2MB of bandwidth is useless, it won't be able to send the 4MB receipt with this grant. This forces the bandwidth scheduler to either give the shard all of the requested bandwidth (the total size of outgoing receipts), or nothing. Otherwise there could be a liveness issue - shards that want to send large receipts, but can't because they're receiving partial bandwidth grants, which aren't enough to send the larger receipt.
That's why bandwidth request is more granular - it says "granting me one of those proposed values of bandwidth is enough for me to send something and make progress. Please grant me one value from the requested list of values, I can make progress with any of them, but being granted more will allow me to send out more receipts." Thanks to this bandwidth scheduler will be able to differentiate between the situation where there are many small receipts (and it can grant less than the total buffer size) and the situation where there is one huge receipt (and it has to grant a lot of bandwidth to send out this receipt).

> And why does this structure need to be part of trie and not something like CongestionInfo (in case it's small)?

Generating the structure from scratch every time would require walking over all of the receipts in the outgoing buffer, and that would be slow. Because of that I want to have a structure which sits in the trie and is updated incrementally as receipts are buffered/unbuffered.

File: core/store/src/trie/outgoing_metadata.rs (hunk 2 above)
Comment (jancionear) [reply core/store/src/trie/outgoing_metadata.rs:138]: It doesn't, but it was convenient to have it there for the methods that add and remove receipts. The migrations aren't really a concern, we can change the threshold any time we want, it is only used for the last group. The first group is removed when its size reaches zero, the threshold does't matter.

File: core/primitives/src/trie_key.rs (hunk 1 above)
Comment (jancionear) [reply core/primitives/src/trie_key.rs:185]: I guess we could have an optimized trie queue which keeps multiple groups in a single trie value, but that sounds even more complex.

File: core/primitives/src/trie_key.rs [hunk 8]
```diff
@@ -337,6 +345,10 @@ impl TrieKey {
                 buf.extend(&index.to_le_bytes());
             }
             TrieKey::BandwidthSchedulerState => buf.push(col::BANDWIDTH_SCHEDULER_STATE),
+            TrieKey::OutgoingBufferReceiptSizes { to_shard } => {
```
[the remark this answers is not in the document]
Comment (jancionear) [reply core/primitives/src/trie_key.rs:348]: Yeah I think the comment is wrong, we serialize all `TrieKeys` as a single byte (two nibbles). It was added by Jakob, I can remove it.

File: core/primitives/src/trie_key.rs (hunk 1 above)
Comment (jancionear) [reply core/primitives/src/trie_key.rs:185]: The framework for writing NEAR contracts has some neat wrappers for common datastructures that need to operate on top of trie, maybe we could add something similar in the internal runtime instead of implementing things manually each time.

Comment (jancionear) [reply core/primitives/src/trie_key.rs:185]: Another bad thing about the current struct is that it's not rollbackable, which is a bit unsettling. But I looked at the uses of `forward_or_buffer_receipt` and it didn't look like we rollback after sending out the receipt at any point. I guess congestion info is also not rollbackable.

Comment (wacban) [reply core/primitives/src/trie_key.rs:185]: I guess if we read all of it every time it doesn't matter that much if it's a single or multiple nodes, if anything multiple makes it worse. Have you considered any other data structures for maintaining aggregate size info about receipts?

Comment (jancionear) [reply core/primitives/src/trie_key.rs:185]: > Have you considered any other data structures for maintaining aggregate size info about receipts?

I couldn't find anything better than this. The problem is that we want to have sizes for some prefixes from 0 to n, and we want these sizes to increase by smallish amounts (~100kB). But if you have any cool ideas I'd love to explore them.

Comment (jancionear) [reply core/primitives/src/trie_key.rs:185]: Well I guess you maybe could maintain something like an AVL tree, and walk down to subtrees until you hit some size, but that sounds super complex.

File: core/primitives/src/bandwidth_scheduler.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (wacban) [reply core/primitives/src/bandwidth_scheduler.rs:140]: Alright but I'm still not a fan, especially in light of dynamic resharding in the future where the shard ids will be out of our control. Can you add a test like this one - `test_shard_id_u16_optimization` - but for u8?

File: core/primitives/src/bandwidth_scheduler.rs [hunk 9]
```diff
@@ -129,6 +129,39 @@ impl BandwidthRequest {
 
         Some(BandwidthRequest { to_shard, requested_values_bitmap: bitmap })
     }
+
+    /// Create a basic bandwidth request when receipt sizes are not available.
+    /// It'll request two values - one corresponding to the first receipt in
+    /// the outgoing buffer and one corresponding to max_receipt_size.
```
[the remark this answers is not in the document]
Comment (jancionear) [reply core/primitives/src/bandwidth_scheduler.rs:135]: max_receipt_size = 4MiB
max_shard_bandwidth = 4.5MB

In the current bandwidth limits `max_shards_bandwidth` is equal to 4.5MB + 100kB * `num_shards`

File: core/store/src/trie/outgoing_metadata.rs [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+use super::TrieAccess;
+
+/// Keeps metadata about receipts stored in the outgoing buffers.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct OutgoingMetadatas {
+    /// Metadata for each shard.
+    pub metadatas: BTreeMap<ShardId, OutgoingBufferMetadata>,
+    /// Receipts are grouped into groups and metadatas keep group-level information.
+    /// A new group is started when the size of the last group exceeds this threshold.
+    pub group_size_threshold: u32,
```
Comment (wacban) [inline core/store/src/trie/outgoing_metadata.rs:21]: nit: Maybe wrap it in a config struct? It's kind of similar to the CongestionControl struct that holds both the info and the config, and actually contains the business logic.

File: runtime/runtime/src/lib.rs [hunk 11]
```diff
@@ -1512,15 +1508,15 @@ impl Runtime {
             &apply_state.config.congestion_control_config,
         );
 
+        receipt_sink.save_outgoing_buffer_metadatas(&mut processing_state.state_update);
+
```
Comment (wacban) [inline runtime/runtime/src/lib.rs:1512]: Sanity check
This here covers the new chunk case. The old chunk is an early returns a few lines above that. Do you not call it there because the metadata does not change for a missing chunk?

File: core/store/src/trie/outgoing_metadata.rs [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+        let metadata = self.metadatas.get_mut(&shard_id).expect("Metadata should exist");
+        metadata.on_receipt_removed(receipt_size);
+    }
+}
+
+/// Keeps metadata about receipts stored in outgoing buffer to some shard.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct OutgoingBufferMetadata {
+    pub receipt_group_sizes: ReceiptGroupSizes,
+}
```
Comment (wacban) [inline core/store/src/trie/outgoing_metadata.rs:88]: This struct seems to be redundant - including the impl - can you clarify why did you add it?

File: core/store/src/trie/outgoing_metadata.rs [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+                state_update,
+                *shard_id,
+                &metadata.receipt_group_sizes,
+            );
+        }
+        state_update.commit(StateChangeCause::SaveOutgoingBufferMetadata);
+    }
+
+    /// Update the metadata when a new receipt is added at the end of the buffer.
+    pub fn on_receipt_buffered(&mut self, shard_id: ShardId, receipt_size: u64) {
```
Comment (wacban) [inline core/store/src/trie/outgoing_metadata.rs:69]: nit: Elsewhere it's called pushed and popped - can you pick one and use it everywhere?

File: core/store/src/trie/outgoing_metadata.rs (hunk 2 above)
Comment (wacban) [reply core/store/src/trie/outgoing_metadata.rs:138]: This is actually held in two places, here and in the `OutgoingMetadatas`, it seems like a stretch.

File: core/store/src/trie/outgoing_metadata.rs [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+    pub group_size_threshold: u32,
+}
+
+impl ReceiptGroupSizesV0 {
+    pub fn new(group_size_threshold: u32) -> Self {
+        Self { groups: VecDeque::new(), group_size_threshold }
+    }
+
+    /// Update the groups when a new receipt is added at the end of the buffer.
+    pub fn on_receipt_pushed(&mut self, receipt_size: u64) {
```
Comment (wacban) [inline core/store/src/trie/outgoing_metadata.rs:147]: Is this using the receipt size that is stored in the StateStoredReceipt? It's important for protocol upgrades if we ever decide to change the size accounting logic.

File: core/store/src/trie/outgoing_metadata.rs [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+    /// Update the groups when a receipt is removed from the front of the buffer.
+    pub fn on_receipt_popped(&mut self, receipt_size: u64) {
+        let receipt_size: u32 = receipt_size_to_u32(receipt_size);
+
+        let first_group = self.groups.front_mut().expect(
+            "OutgoingBufferMetadataV0::on_receipt_popped fatal error - there are no groups",
+        );
+
+        first_group.group_size =
+            first_group.group_size.checked_sub(receipt_size).unwrap_or_else(|| {
```
Comment (wacban) [inline core/store/src/trie/outgoing_metadata.rs:180]: nit: why not expect?

File: core/store/src/trie/receipts_column_helper.rs [hunk 16]
```diff
@@ -232,6 +232,35 @@ impl ShardsOutgoingReceiptBuffer {
...
+        };
+
+        if indices.len() == 0 {
+            return Ok(None);
+        }
+
+        let receipt_key =
+            TrieKey::BufferedReceipt { receiving_shard: to_shard, index: indices.first_index };
+        let trie_receipt =
+            if side_effects { get(trie, &receipt_key) } else { get_pure(trie, &receipt_key) };
```
Comment (wacban) [inline core/store/src/trie/receipts_column_helper.rs:255]: oh my

Comment (wacban) [reply core/store/src/trie/receipts_column_helper.rs:255]: this is really easy to forget, we should make side_effects an argument to get instead

File: core/store/src/trie/receipts_column_helper.rs [hunk 17]
```diff
@@ -232,6 +232,35 @@ impl ShardsOutgoingReceiptBuffer {
...
     }
+
+    /// Fetch the first receipt from the buffer.
+    /// Can't implement using OutgoingReceiptBuffer::iter() because it requires a &mut reference.
+    pub fn get_first_receipt(
+        &self,
+        to_shard: ShardId,
+        trie: &dyn TrieAccess,
+        side_effects: bool,
+    ) -> Result<Option<ReceiptOrStateStoredReceipt>, StorageError> {
```
Comment (wacban) [inline core/store/src/trie/receipts_column_helper.rs:243]: Does this method actually remove the receipt from the buffer? Does it clone it or will it be read again later? Do we care?

Comment (wacban) [reply core/store/src/trie/receipts_column_helper.rs:243]: nit: maybe call it peek_receipt?

File: runtime/runtime/src/congestion_control.rs [hunk 18]
```diff
@@ -335,9 +408,97 @@ impl ReceiptSinkV2<'_> {
         self.own_congestion_info.add_receipt_bytes(size)?;
         self.own_congestion_info.add_buffered_receipt_gas(gas)?;
 
+        if receipt.queue_metadata_version() == Some(QueueMetadataVersion::OutgoingBufferMetadataV0)
+        {
+            self.outgoing_buffer_metadatas.on_receipt_buffered(shard, size);
+        }
```
Comment (wacban) [inline runtime/runtime/src/congestion_control.rs:414]: nit: Maybe make it into a match so that when new versions are added the compiler will remind us to handle it.

File: runtime/runtime/src/congestion_control.rs [hunk 19]
```diff
@@ -335,9 +408,97 @@ impl ReceiptSinkV2<'_> {
...
+            // Collect sizes of receipts stored in this outgoing buffer.
+            // We need to start with the size of the first receipt. It could happen that
+            // the size of the first group is larger than max_receipt_size, and it could be that
+            // num_shards * base_bandwidth + first_group_size > max_shard_bandwidth, in which case
+            // we would never be able to grant the bandwidth required to send the first group.
+            // It's an unlikely scenario, but to protect against it we always request bandwidth for the first receipt,
+            // it's always possible to grant enough bandwidth for one receipt.
+            let mut receipt_group_sizes = metadata.iter_receipt_group_sizes();
+            let first_group_size = receipt_group_sizes.next().unwrap();
+            let receipt_sizes = [first_receipt_size, first_group_size - first_receipt_size]
```
Comment (wacban) [inline runtime/runtime/src/congestion_control.rs:490]: nit: It's a bit strange that the bandwidth request is made based on receipt sizes but you pass in the group sizes. Maybe make it be based on groups and just pass the first receipt size as an additional argument to the ctor? I'm not sure if this suggestion is any good, feel free to ignore.

**Review by wacban — COMMENTED** [review]
I can't say I understand all of it but I did read it and it makes sense. Added some comments. Let me think a bit about how this will work with resharding.

**Comment by jancionear** [comment]
I tried out how converting the receipt groups to a TrieQueue would look like, you can check it out in this PR: https://github.com/near/nearcore/pull/12410

I have mixed feelings. It's nice that it's rollback-friendly and there's no risk of having a super large value, but the code got way more complex and ugly :/

File: core/store/src/trie/outgoing_metadata.rs (hunk 2 above)
Comment (jancionear) [reply core/store/src/trie/outgoing_metadata.rs:138]: It needs to be in `OutgoingMetadatas` because I have to pass it to the constructor of some shard's metadata when a new metadata is created. So there are at least two places - `OutgoingMetadatas` and `OutgoingBufferMetadata` where it needs to be.

File: core/primitives/src/receipt.rs (hunk 3 above)
Comment (jancionear) [reply core/primitives/src/receipt.rs:124]: Alright. Initially I used the version of stored receipt, but then I thought that it could be clearer to mark the version of the metadata explicitly. I'll remove it.

File: core/primitives/src/bandwidth_scheduler.rs [hunk 20]
```diff
@@ -129,6 +129,39 @@ impl BandwidthRequest {
...
+                    break;
+                }
+            }
+        }
+
+        for i in 0..values.len() {
+            if values[i] == params.max_receipt_size {
+                bitmap.set_bit(i, true);
+                break;
+            }
```
[the remark this answers is not in the document]
Comment (jancionear) [reply core/primitives/src/bandwidth_scheduler.rs:160]: Yes it's a guarantee. `BandwidthRequestValues::new` generates values and then replaces the value closes to `max_receipt_size` with `max_receipt_size`.
It's needed because there's only a guarantee that:
`base_bandwidth * num_shards + max_receipt_size <= max_shard_bandwidth`
But there's no such guarantee for the first value that is larger than `max_shard_bandwidth`.

I'll add an assert here, as it's an important invariant.

File: core/primitives/src/bandwidth_scheduler.rs [hunk 21]
```diff
@@ -129,6 +129,39 @@ impl BandwidthRequest {
...
         Some(BandwidthRequest { to_shard, requested_values_bitmap: bitmap })
     }
+
+    /// Create a basic bandwidth request when receipt sizes are not available.
+    /// It'll request two values - one corresponding to the first receipt in
+    /// the outgoing buffer and one corresponding to max_receipt_size.
+    /// This ensures that the blockchain will be able to make progress while
+    /// waiting for the receipt sizes to become available.
+    /// The resulting behaviour is similar to the previous approach with allowed shard.
+    pub fn make_basic(
```
[the remark this answers is not in the document]
Comment (jancionear) [reply core/primitives/src/bandwidth_scheduler.rs:139]: Yes, when there are no receipts `generate_bandwidth_request` returns `None`
```rust
let Some(first_receipt) =
            self.outgoing_buffers.get_first_receipt(to_shard, trie, side_effects)?
        else {
            // No receipts in the outgoing buffer
            return Ok(None);
        };
```

File: core/store/src/trie/outgoing_metadata.rs [hunk 22]
```diff
@@ -0,0 +1,212 @@
...
+use near_schema_checker_lib::ProtocolSchema;
+use near_vm_runner::logic::ProtocolVersion;
+
+use crate::TrieUpdate;
+
+use super::TrieAccess;
+
+/// Keeps metadata about receipts stored in the outgoing buffers.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct OutgoingMetadatas {
```
[the remark this answers is not in the document]
Comment (jancionear) [reply core/store/src/trie/outgoing_metadata.rs:16]: Yeah it's a bit annoying, I'm tempted to name everything `OutgoingBuffered`

File: core/store/src/trie/outgoing_metadata.rs (hunk 12 above)
Comment (jancionear) [reply core/store/src/trie/outgoing_metadata.rs:88]: I added it as a sort of interface for the outgoing metadata for some shard. It's something that can be loaded, updated when a receipt is added/removed, and provides some information about the receipts. It doesn't matter what the internal implementation is.
This struct could be `pub`, while `ReceiptGroupSizes` could be private

File: core/store/src/trie/outgoing_metadata.rs (hunk 14 above)
Comment (jancionear) [reply core/store/src/trie/outgoing_metadata.rs:147]: Yes, it's the same size as in congestion control

File: core/store/src/trie/receipts_column_helper.rs (hunk 17 above)
Comment (jancionear) [reply core/store/src/trie/receipts_column_helper.rs:243]: It doesn't remove anything. In Rust `first()` and `get()` are usually methods that access a collection without removing the item, so I thought that this would be similar.
I can change it to peek if it's confusing

File: runtime/runtime/src/congestion_control.rs (hunk 18 above)
Comment (jancionear) [reply runtime/runtime/src/congestion_control.rs:414]: Good point, will do :+1:

File: runtime/runtime/src/congestion_control.rs (hunk 19 above)
Comment (jancionear) [reply runtime/runtime/src/congestion_control.rs:490]: BandwidthRequest would ideally operate on receipts, but groups are an optimization that has to be done. But it's a bit unexact, I'll add a TODO and revisit it later, there's enough going on in this PR.

File: core/primitives/src/trie_key.rs (hunk 1 above)
Comment (shreyan-gupta) [reply core/primitives/src/trie_key.rs:185]: I have a couple of ideas I'd like to discuss in the meeting

File: core/store/src/trie/outgoing_metadata.rs (hunk 10 above)
Comment (shreyan-gupta) [reply core/store/src/trie/outgoing_metadata.rs:21]: Agree

File: runtime/runtime/src/congestion_control.rs (hunk 19 above)
Comment (jancionear) [reply runtime/runtime/src/congestion_control.rs:490]: I tried to find a better name, but it's pretty hard. Bandwidth request can be created from receipts or receipt groups, or prefix sums of receipt sizes. `receipt_sizes` is generic enough to cover all these cases.
If you have an idea for a better name, I can change it.

**Comment by jancionear** [comment]
Closing in favor of https://github.com/near/nearcore/pull/12464
